inc/saiarsqualitymap.h

```diff
@@ -0,0 +1,353 @@
+ *    assistance with these files: Intel Corporation, Mellanox Technologies Ltd,
+ *    Dell Products, L.P., Facebook, Inc., Marvell International Ltd.
+ *
+ * @file    saiarsqualitymap.h
+ *
+ * @brief   This module defines SAI QOS Maps interface
+ */
+
+#if !defined (__SAIARSQUALITYMAP_H_)
+#define __SAIARSQUALITYMAP_H_
+
+#include <saitypes.h>
+
+/**
+ * @defgroup SAIARSQUALITYMAP SAI - ARS quality map specific API definitions.
+ *
+ * @{
+ */
+
+/**
+ * @brief Adaptive Routing and Switching quality measure algorithm
+ */
+typedef enum _sai_ars_quality_map_algo_t
+{
+    /** Exponentially weighted moving average */
+    SAI_ARS_QUALITY_MAP_ALGO_EWMA,
+
+} sai_ars_quality_map_algo_t;
+
+/**
+ * @brief Enum defining attributes for ARS quality map.
+ */
+typedef enum _sai_ars_quality_map_attr_t
+{
+    /**
+     * @brief Start of attributes
+     */
+    SAI_ARS_QUALITY_MAP_ATTR_START,
+
+    /**
+     * @brief Sampling interval in microseconds of data for quality measure computation
+     *
+     * @type sai_uint32_t
+     * @flags CREATE_AND_SET
+     * @default 16
+     */
+    SAI_ARS_QUALITY_MAP_ATTR_SAMPLING_INTERVAL = SAI_ARS_QUALITY_MAP_ATTR_START,
+
+    /**
+     * @brief Enable past port load as the quality parameter. This is the average egress bytes measured on a port
+     *
+     * @type bool
+     * @flags CREATE_AND_SET
+     * @default true
+     */
+    SAI_ARS_QUALITY_MAP_ATTR_PORT_LOAD_PAST,
+
+    /**
+     * @brief Weight attribute is used in EWMA calculations.
+     * Large weight lowers the significance of instantaneous measure on the overall average.
+     *
+     * @type sai_uint8_t
+     * @flags CREATE_AND_SET
+     * @default 16
+     */
+    SAI_ARS_QUALITY_MAP_ATTR_PORT_LOAD_PAST_WEIGHT,
+
+    /**
+     * @brief Enable future port load as the quality parameter. This is the average queued bytes measured on a port.
+     *
+     * @type bool
+     * @flags CREATE_AND_SET
+     * @default true
+     */
+    SAI_ARS_QUALITY_MAP_ATTR_PORT_LOAD_FUTURE,
+
+    /**
+     * @brief Weight attribute is used in EWMA calculations.
+     * Large weight lowers the significance of instantaneous measure on the overall average.
+     *
+     * @type sai_uint8_t
+     * @flags CREATE_AND_SET
+     * @default 16
+     */
+    SAI_ARS_QUALITY_MAP_ATTR_PORT_LOAD_FUTURE_WEIGHT,
+
+    /**
+     * @brief Set port loading value to current sampled value when sampled value is less than the average.
+     *
+     * @type bool
+     * @flags CREATE_AND_SET
+     * @default false
+     */
+    SAI_ARS_QUALITY_MAP_ATTR_PORT_LOAD_CURRENT,
+
+    /**
+     * @brief EWMA exponent used in port loading computation. Larger the exponent, larger is the weight to previously computed port loading value.
+     *
+     * @type sai_uint8_t
+     * @flags CREATE_AND_SET
+     * @default 2
+     */
+    SAI_ARS_QUALITY_MAP_ATTR_PORT_LOAD_EXPONENT,
+
+    /**
+     * @brief Number of quantization bands support for quality map.
+     *
```

Comment:
Can we consider the additional params to be used for port grading ? For example, buffer/queue-based in additional to bandwidth-based based (link utlization like in your case) ? So, both parameters can impact the port grading for choosing it by ARS algo

Reply:
Additional parameters can be added as need by other vendors. For now I have added some as a starting point for a given HW support.